packages/cli/src/util/dev/server.ts

```diff
@@ -1430,37 +1406,144 @@ export default class DevServer {
-        const origUrl = url.parse(req.url || '/', true);
-        origUrl.pathname = middlewareResult.pathname;
-        prevUrl = url.format(origUrl);
-      }
-      if (middlewareResult.query && prevUrl) {
-        const origUrl = url.parse(req.url || '/', true);
-        delete origUrl.search;
-        Object.assign(origUrl.query, middlewareResult.query);
-        prevUrl = url.format(origUrl);
+      if (startMiddlewareResult) {
+        const { port, pid } = startMiddlewareResult;
+        this.devServerPids.add(pid);
+
+        const middlewareReqHeaders = nodeHeadersToFetchHeaders(req.headers);
+
+        // Add the Vercel platform proxy request headers
+        const proxyHeaders = this.getProxyHeaders(req, requestId, true);
+        for (const [name, value] of nodeHeadersToFetchHeaders(proxyHeaders)) {
+          middlewareReqHeaders.set(name, value);
+        }
+
+        try {
+          const middlewareRes = await fetch(
+            `http://127.0.0.1:${port}${parsed.path}`,
+            {
+              headers: middlewareReqHeaders,
+              method: req.method,
+              redirect: 'manual',
+            }
+          );
+
+          // Apply status code from middleware invocation,
+          // for i.e. redirects or a custom 404 page
+          res.statusCode = middlewareRes.status;
+
+          let rewritePath = '';
+          let contentType = '';
+          let shouldContinue = false;
+          const skipMiddlewareHeaders = new Set([
+            'date',
+            'connection',
+            'content-length',
+            'transfer-encoding',
+          ]);
+          for (const [name, value] of middlewareRes.headers) {
+            if (name === 'x-middleware-next') {
+              shouldContinue = value === '1';
+            } else if (name === 'x-middleware-rewrite') {
+              rewritePath = value;
+              shouldContinue = true;
+            } else if (name === 'content-type') {
+              contentType = value;
+            } else if (!skipMiddlewareHeaders.has(name)) {
+              // Any other kind of response header should be included
+              // on both the incoming HTTP request (for when proxying
+              // to another function) and the outgoing HTTP response.
+              res.setHeader(name, value);
+              req.headers[name] = value;
+            }
+          }
+
+          if (!shouldContinue) {
+            const middlewareBody = await middlewareRes.buffer();
+            this.setResponseHeaders(res, requestId);
+            if (middlewareBody.length > 0) {
+              res.setHeader('content-length', middlewareBody.length);
+              if (contentType) {
+                res.setHeader('content-type', contentType);
+              }
+              res.end(middlewareBody);
+            } else {
+              res.end();
+            }
+            return;
+          }
+
+          if (rewritePath) {
+            // TODO: add validation?
```

Comment:
Should we make a card for this?

Reply:
@TooTallNate How's this?
https://linear.app/vercel/issue/VCCLI-153/validate-rewritepath-before-using-in-middleware-during-vc-dev